consensus/hotstuff/timeoutcollector/aggregation.go

```diff
@@ -69,43 +79,33 @@ func NewWeightedMultiMessageSigAggregator(ids flow.IdentityList, // list of all
 	return &TimeoutSignatureAggregator{
 		hasher:        crypto.NewBLSKMAC(dsTag),
 		idToInfo:      idToInfo,
-		idToSignature: make(map[flow.Identifier]crypto.Signature),
+		idToSignature: make(map[flow.Identifier]sigInfo),
+		view:          view,
 	}, nil
 }
 
-// Verify verifies the signature under the stored public keys.
+// VerifyAndAdd verifies the signature under the stored public keys and adds signature with corresponding
+// highest QC view to the internal set. Internal set and collected weight is modified iff signature _is_ valid.
+// The total weight of all collected signatures (excluding duplicates) is returned regardless
+// of any returned error.
 // Expected errors during normal operations:
 //  - model.InvalidSignerError if signerID is invalid (not a consensus participant)
+//  - model.DuplicatedSignerError if the signer has been already added
 //  - model.ErrInvalidSignature if signerID is valid but signature is cryptographically invalid
 // The function is thread-safe.
-func (a *TimeoutSignatureAggregator) Verify(signerID flow.Identifier, sig crypto.Signature, msg []byte) error {
+func (a *TimeoutSignatureAggregator) VerifyAndAdd(signerID flow.Identifier, sig crypto.Signature, highestQCView uint64) (totalWeight uint64, exception error) {
```

Comment:
It is better to verify the signature cryptographically only when the signer ID is not duplicate, because the cryptographic verification is more expensive.

Reply:
Makes sense, will change this